### PR TITLE
Move to custom exception types

### PR DESCRIPTION
**Pull request by kashike**
Opening this to get some initial feedback before continuing.

Rather than using a single type (`IllegalStateException`) when notifying consumers about exceptions, use custom ones - this makes catching and handling exceptions easier.

### CODE REVIEW

**Comment by kashike** [comment]
@chrisvest When you have a minute, please take a look and see what you think before I continue. Started this before the comments in #70.

**Comment by chrisvest** [comment]
@kashike I like the idea. I think the exceptions should be defined in each their own class, though.

File: buffer-api/src/main/java/io/netty/buffer/api/internal/ResourceSupport.java [hunk 1]
```diff
@@ -52,7 +53,7 @@ protected ResourceSupport(Drop<T> drop) {
      */
     protected final I acquire() {
         if (acquires < 0) {
-            throw attachTrace(new IllegalStateException("This resource is closed: " + this + '.'));
+            throw attachTrace(new BufferClosedException("This resource is closed: " + this + '.'));
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/internal/ResourceSupport.java]: `ResourceSupport` should not be buffer specific. We can delegate the exception creation to a protected abstract method that subclasses override. Then we can also make the exception message be "This _buffer_ is closed…"

File: buffer-api/src/main/java/io/netty/buffer/api/Buffer.java [hunk 2]
```diff
@@ -377,7 +378,9 @@ default ByteCursor openReverseCursor() {
      * {@code false}.
      *
      * @param size The requested number of bytes of space that should be available for writing.
-     * @throws IllegalStateException if this buffer is not in a bad state, or is {@linkplain #readOnly() read-only}.
+     * @throws IllegalStateException if this buffer is not in a bad state.
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/Buffer.java]: Remove "not" – makes the rebase easier when #72 and this eventually conflicts.

File: buffer-api/src/main/java/io/netty/buffer/api/Buffer.java [hunk 3]
```diff
@@ -418,7 +421,9 @@ default void ensureWritable(int size) {
      * @param allowCompaction {@code true} if the method is allowed to modify the
      *                                   {@linkplain #readerOffset() reader offset} and
      *                                   {@linkplain #writerOffset() writer offset}, otherwise {@code false}.
-     * @throws IllegalStateException if this buffer is not in a bad state, or is {@linkplain #readOnly() read-only}.
+     * @throws BufferReadOnlyException if this buffer is {@linkplain #readOnly() read-only}.
+     * @throws IllegalArgumentException if {@code size} or {@code minimumGrowth} are negative.
+     * @throws IllegalStateException if this buffer is not in a bad state.
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/Buffer.java]: Remove "not" – makes the rebase easier when #72 and this eventually conflicts.

File: buffer-api/src/main/java/io/netty/buffer/api/Buffer.java [hunk 4]
```diff
@@ -553,7 +558,8 @@ default Buffer split() {
     /**
      * Discards the read bytes, and moves the buffer contents to the beginning of the buffer.
      *
-     * @throws IllegalStateException if this buffer is not in a bad state, or is {@linkplain #readOnly() read-only}.
+     * @throws BufferReadOnlyException if this buffer is {@linkplain #readOnly() read-only}.
+     * @throws IllegalStateException if this buffer is not in a bad state.
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/Buffer.java]: Remove "not" – makes the rebase easier when #72 and this eventually conflicts.

**Review by chrisvest — CHANGES_REQUESTED** [review]

**Comment by chrisvest** [comment]
Also, I've no idea why test report uploading keeps failing on your PRs but works on mine 🤷

**Comment by normanmaurer** [comment]
@chrisvest I wonder if the uploading this is some sort of a permission problem. I had the same problem in netty at some point and so switched to the extra workflow for it:

https://github.com/netty/netty/blob/4.1/.github/workflows/ci-pr-reports.yml

**Comment by chrisvest** [comment]
@kashike I merged a separate workflow for uploading test results, so if you rebase _maybe_ that part will work now.

File: buffer-api/src/main/java/io/netty/buffer/api/BufferClosedException.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+/**
+ * An exception thrown when an operation is attempted on a {@link Buffer} when it has been closed.
+ */
+public class BufferClosedException extends IllegalStateException {
```
Comment (normanmaurer) [inline buffer-api/src/main/java/io/netty/buffer/api/BufferClosedException.java]: I think I would mark all our exceptions as final for now.

File: buffer-api/src/main/java/io/netty/buffer/api/BufferReadOnlyException.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+/**
+ * An exception thrown when an operation is attempted on a {@linkplain Buffer#readOnly() read-only} {@link Buffer}.
+ */
+public class BufferReadOnlyException extends IllegalStateException {
```
Comment (normanmaurer) [inline buffer-api/src/main/java/io/netty/buffer/api/BufferReadOnlyException.java]: Why we don't just use what the JDK provides here ? 

https://docs.oracle.com/javase/7/docs/api/java/nio/ReadOnlyBufferException.html

At least I would extend what the JDK provides as `UnsupportedOperationException` sounds also more correct.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer-api/src/main/java/io/netty/buffer/api/BufferReadOnlyException.java (hunk 6 above)
Comment (kashike) [reply buffer-api/src/main/java/io/netty/buffer/api/BufferReadOnlyException.java]: The JDK class does not allow specifying a message.

Comment (kashike) [reply buffer-api/src/main/java/io/netty/buffer/api/BufferReadOnlyException.java]: I have changed the parent of ours to `UnsupportedOperationException`.

**Comment by kashike** [comment]
I've addressed the comments left earlier.

**Comment by kashike** [comment]
I have had to do some... ugly things to the tests. Due to the exceptions now having different parents, it becomes a mess. Perhaps we should introduce a root exception type (`BufferException extends RuntimeException`) for our exceptions to extend?

We could then also introduce more exception types for what we throw.

File: buffer-api/src/main/java/io/netty/buffer/api/internal/Statics.java [hunk 7]
```diff
@@ -165,12 +167,12 @@ private static void bbputFallback(ByteBuffer dest, int destPos, ByteBuffer src,
...
-    static IllegalStateException bufferIsClosed() {
-        return new IllegalStateException("This buffer is closed.");
+    static BufferClosedException bufferIsClosed(Buffer buffer) {
+        return new BufferClosedException("This buffer is closed.");
     }
 
-    static IllegalStateException bufferIsReadOnly() {
-        return new IllegalStateException("This buffer is read-only.");
+    static BufferReadOnlyException bufferIsReadOnly(Buffer buffer) {
+        return new BufferReadOnlyException("This buffer is read-only.");
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/internal/Statics.java]: Why do these two take `Buffer` as arguments now?

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferComponentIterationTest.java [hunk 8]
```diff
@@ -357,7 +359,7 @@ public void forEachWritableOnClosedBufferMustThrow(Fixture fixture) {
         try (BufferAllocator allocator = fixture.createAllocator()) {
             Buffer buf = allocator.allocate(8);
             buf.close();
-            assertThrows(IllegalStateException.class, () -> buf.forEachWritable(0, (index, component) -> true));
+            assertThrows(UnsupportedOperationException.class, () -> buf.forEachWritable(0, (index, component) -> true));
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferComponentIterationTest.java]: Why is this not `BufferClosedException`?

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferCompositionTest.java [hunk 9]
```diff
@@ -391,7 +392,8 @@ public void composingReadOnlyBuffersMustCreateReadOnlyCompositeBuffer() {
              Buffer b = allocator.allocate(4).makeReadOnly();
              Buffer composite = CompositeBuffer.compose(allocator, a.send(), b.send())) {
             assertTrue(composite.readOnly());
-            verifyWriteInaccessible(composite);
+            verifyWriteInaccessible(composite, BufferReadOnlyException.class, BufferReadOnlyException.class
+            );
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferCompositionTest.java]: Stray line break

File: buffer-api/src/main/java/io/netty/buffer/api/internal/ResourceSupport.java [hunk 10]
```diff
@@ -15,6 +15,7 @@
  */
 package io.netty.buffer.api.internal;
 
+import io.netty.buffer.api.BufferClosedException;
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/internal/ResourceSupport.java]: This import can be removed now.

File: buffer-api/src/main/java/io/netty/buffer/api/adaptor/ByteBufAdaptor.java [hunk 11]
```diff
@@ -253,7 +253,7 @@ public boolean getBoolean(int index) {
     public byte getByte(int index) {
         try {
             return buffer.getByte(index);
-        } catch (IllegalStateException e) {
+        } catch (IllegalStateException | UnsupportedOperationException e) {
```
Comment (chrisvest) [inline buffer-api/src/main/java/io/netty/buffer/api/adaptor/ByteBufAdaptor.java]: I think all of these should just catch `BufferClosedException` and let any `BufferReadOnlyException` bubble through. Buffers being read-only is not a reference count issue. The illegal state probably is, though, so let's keep that.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java [hunk 12]
```diff
@@ -121,7 +122,8 @@ public void compactOnReadOnlyBufferMustThrow(Fixture fixture) {
         try (BufferAllocator allocator = fixture.createAllocator();
              Buffer buf = allocator.allocate(8)) {
             buf.makeReadOnly();
-            assertThrows(IllegalStateException.class, () -> buf.compact());
+            final RuntimeException c1e = assertThrows(RuntimeException.class, () -> buf.compact());
+            assertTrue(c1e instanceof IllegalStateException || c1e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: We should assert that this specifically throws `BufferReadOnlyException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java [hunk 13]
```diff
@@ -215,19 +217,22 @@ public void compactOnConstBufferMustNotImpactSiblings(Fixture fixture) {
                  Buffer c = a.copy()) {
                 assertEquals(1, a.readByte());
                 assertEquals(2, a.readByte());
-                assertThrows(IllegalStateException.class, () -> a.compact()); // Can't compact read-only buffer.
+                final RuntimeException c1e = assertThrows(RuntimeException.class, () -> a.compact()); // Can't compact read-only buffer.
+                assertTrue(c1e instanceof IllegalStateException || c1e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: These should also assert on `BufferReadOnlyException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java [hunk 14]
```diff
@@ -215,19 +217,22 @@ public void compactOnConstBufferMustNotImpactSiblings(Fixture fixture) {
...
                 assertEquals(3, a.readByte());
                 assertEquals(4, a.readByte());
 
                 assertEquals(1, b.readByte());
                 assertEquals(2, b.readByte());
-                assertThrows(IllegalStateException.class, () -> b.compact()); // Can't compact read-only buffer.
+                final RuntimeException c2e = assertThrows(RuntimeException.class, () -> b.compact()); // Can't compact read-only buffer.
+                assertTrue(c2e instanceof IllegalStateException || c2e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: 📌

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java [hunk 15]
```diff
@@ -215,19 +217,22 @@ public void compactOnConstBufferMustNotImpactSiblings(Fixture fixture) {
...
                 assertEquals(3, b.readByte());
                 assertEquals(4, b.readByte());
 
                 assertEquals(1, c.readByte());
                 assertEquals(2, c.readByte());
-                assertThrows(IllegalStateException.class, () -> c.compact()); // Can't compact read-only buffer.
+                final RuntimeException c3e = assertThrows(RuntimeException.class, () -> c.compact()); // Can't compact read-only buffer.
+                assertTrue(c3e instanceof IllegalStateException || c3e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: 📌

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReferenceCountingTest.java [hunk 16]
```diff
@@ -77,7 +77,7 @@ void acquireOnClosedBufferMustThrow(Fixture fixture) {
         try (BufferAllocator allocator = fixture.createAllocator()) {
             var buf = allocator.allocate(8);
             buf.close();
-            assertThrows(IllegalStateException.class, () -> acquire((ResourceSupport<?, ?>) buf));
+            assertThrows(UnsupportedOperationException.class, () -> acquire((ResourceSupport<?, ?>) buf));
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReferenceCountingTest.java]: This should assert on `BufferClosedException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java [hunk 17]
```diff
@@ -323,88 +324,91 @@ static void stopExecutor() throws IOException {
     public static void verifyInaccessible(Buffer buf) {
         verifyReadInaccessible(buf);
 
-        verifyWriteInaccessible(buf);
+        verifyWriteInaccessible(buf, BufferClosedException.class, IllegalStateException.class);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java]: This is shining light on a bug in `ensureWritable`. It should throw `BufferClosedException` in these cases, and we can then collapse this to one parameter.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java [hunk 18]
```diff
@@ -121,7 +122,8 @@ public void compactOnReadOnlyBufferMustThrow(Fixture fixture) {
         try (BufferAllocator allocator = fixture.createAllocator();
              Buffer buf = allocator.allocate(8)) {
             buf.makeReadOnly();
-            assertThrows(IllegalStateException.class, () -> buf.compact());
+            final RuntimeException c1e = assertThrows(RuntimeException.class, () -> buf.compact());
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: This needs to assert on `BufferReadOnlyException` and `NioBuffer` needs to comply.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java (hunk 13 above)
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: Same here. `NioBuffer` is wrong for diverging.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java (hunk 14 above)
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: 📌

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java (hunk 15 above)
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: 📌

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java [hunk 19]
```diff
@@ -37,7 +37,8 @@ public void closingBufRefMustCloseOwnedBuf() {
             ref.contents().writeInt(42);
             assertThat(ref.contents().readInt()).isEqualTo(42);
             ref.close();
-            assertThrows(IllegalStateException.class, () -> ref.contents().writeInt(32));
+            final RuntimeException w1e = assertThrows(RuntimeException.class, () -> ref.contents().writeInt(32));
+            assertTrue(w1e instanceof IllegalStateException || w1e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java]: This should assert on `BufferClosedException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java [hunk 20]
```diff
@@ -49,7 +50,7 @@ public void closingBufRefMustCloseOwnedBufFromSend() {
             ref.contents().writeInt(42);
             assertThat(ref.contents().readInt()).isEqualTo(42);
             ref.close();
-            assertThrows(IllegalStateException.class, () -> ref.contents().writeInt(32));
+            assertThrows(UnsupportedOperationException.class, () -> ref.contents().writeInt(32));
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java]: This should be `BufferClosedException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java [hunk 21]
```diff
@@ -72,11 +73,13 @@ public void mustCloseOwnedBufferWhenReplaced() {
                 ref.replace(buf.send()); // Pass replacement directly.
             }
 
-            assertThrows(IllegalStateException.class, () -> orig.get().writeInt(32));
+            final RuntimeException w1e = assertThrows(RuntimeException.class, () -> orig.get().writeInt(32));
+            assertTrue(w1e instanceof IllegalStateException || w1e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java]: These two should assert on `BufferClosedException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java [hunk 22]
```diff
@@ -72,11 +73,13 @@ public void mustCloseOwnedBufferWhenReplaced() {
...
             ref.contents().writeInt(42);
             assertThat(ref.contents().readInt()).isEqualTo(42);
             ref.close();
-            assertThrows(IllegalStateException.class, () -> ref.contents().writeInt(32));
+            final RuntimeException w2e = assertThrows(RuntimeException.class, () -> ref.contents().writeInt(32));
+            assertTrue(w2e instanceof IllegalStateException || w2e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java]: 📌

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java [hunk 23]
```diff
@@ -99,11 +102,13 @@ public void mustCloseOwnedBufferWhenReplacedFromSend() {
...
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java]: Same for these two. Looks like this test got duplicated somehow (the one above this one is the duplicate, which can be removed).

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java [hunk 24]
```diff
@@ -113,7 +118,8 @@ public void sendingRefMustSendBuffer() {
              BufferRef refA = new BufferRef(allocator.allocate(8).send())) {
             refA.contents().writeInt(42);
             Send<BufferRef> send = refA.send();
-            assertThrows(IllegalStateException.class, () -> refA.contents().readInt());
+            final RuntimeException r1e = assertThrows(RuntimeException.class, () -> refA.contents().readInt());
+            assertTrue(r1e instanceof IllegalStateException || r1e instanceof UnsupportedOperationException);
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferRefTest.java]: This should assert on `BufferClosedException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java [hunk 25]
```diff
@@ -323,88 +324,91 @@ static void stopExecutor() throws IOException {
...
 
         try (BufferAllocator allocator = BufferAllocator.heap();
              Buffer target = allocator.allocate(24)) {
-            assertThrows(IllegalStateException.class, () -> buf.copyInto(0, target, 0, 1));
-            assertThrows(IllegalStateException.class, () -> buf.copyInto(0, new byte[1], 0, 1));
-            assertThrows(IllegalStateException.class, () -> buf.copyInto(0, ByteBuffer.allocate(1), 0, 1));
+            assertThrows(UnsupportedOperationException.class, () -> buf.copyInto(0, target, 0, 1));
+            assertThrows(UnsupportedOperationException.class, () -> buf.copyInto(0, new byte[1], 0, 1));
+            assertThrows(UnsupportedOperationException.class, () -> buf.copyInto(0, ByteBuffer.allocate(1), 0, 1));
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java]: These `UnsupportedOperationException` can be `BufferClosedException`.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java [hunk 26]
```diff
@@ -323,88 +324,91 @@ static void stopExecutor() throws IOException {
...
-        assertThrows(IllegalStateException.class, () -> buf.openCursor(0, 0));
-        assertThrows(IllegalStateException.class, () -> buf.openReverseCursor());
-        assertThrows(IllegalStateException.class, () -> buf.openReverseCursor(0, 0));
+        assertThrows(UnsupportedOperationException.class, () -> acquire((ResourceSupport<?, ?>) buf));
+        final RuntimeException c1e = assertThrows(RuntimeException.class, () -> buf.copy());
+        assertTrue(c1e instanceof UnsupportedOperationException || c1e instanceof IllegalStateException);
+        assertThrows(UnsupportedOperationException.class, () -> buf.openCursor());
+        assertThrows(UnsupportedOperationException.class, () -> buf.openCursor(0, 0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.openReverseCursor());
+        assertThrows(UnsupportedOperationException.class, () -> buf.openReverseCursor(0, 0));
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java]: All of these should be `BufferClosedException` as well.

The `split` and `send` methods will need some `isAccessible` checks before their `isOwned` checks.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java [hunk 27]
```diff
@@ -323,88 +324,91 @@ static void stopExecutor() throws IOException {
...
+        assertThrows(UnsupportedOperationException.class, () -> buf.getChar(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getShort(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getUnsignedShort(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getMedium(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getUnsignedMedium(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getInt(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getUnsignedInt(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getFloat(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getLong(0));
+        assertThrows(UnsupportedOperationException.class, () -> buf.getDouble(0));
```
Comment (chrisvest) [inline buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferTestSupport.java]: All of these also need to assert on `BufferClosedException` – pure read methods cannot throw `BufferReadOnlyException`.

**Review by chrisvest — CHANGES_REQUESTED** [review]
Thanks for doing this work. I had a bunch of comments. Please take a look.

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java (hunk 12 above)
Comment (kashike) [reply buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: They don't just throw this, which is why it was changed. 
> org.opentest4j.AssertionFailedError: Unexpected exception type thrown ==> expected: <io.netty.util.IllegalReferenceCountException> but was: <io.netty.buffer.api.BufferClosedException>

File: buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java (hunk 18 above)
Comment (kashike) [reply buffer-tests/src/test/java/io/netty/buffer/api/tests/BufferReadOnlyTest.java]: What would you prefer, moving the read-only check above the owned check?

**Comment by kashike** [comment]
I've pushed some of the requested changes, with around 200 tests failing locally.

**Review by chrisvest — APPROVED** [review]
I fixed the remaining issues

**Comment by kashike** [comment]
Cheers @chrisvest - I was not sure what you had in mind.

**Pull request merged**